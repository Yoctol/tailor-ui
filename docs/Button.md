---
id: button
title: Button
---

To trigger an operation.

## When To Use

A button means an operation (or a series of operations). Clicking a button will trigger corresponding business logic.

## Examples

```js
import { Button } from 'tailor-ui';
```

### Basic

```jsx live
<>
  <Button>Basic</Button>
  <Button rounded ml="2">
    Basic
  </Button>
  <Button disabled ml="2">
    Basic
  </Button>
  <Button loading ml="2">
    Basic
  </Button>
</>
```

### Variant

```jsx live
<>
  <Button variant="primary">Primary</Button>
  <Button variant="primary-invert" ml="2">
    Primary Invert
  </Button>
  <br />
  <br />
  <Button variant="danger">Danger</Button>
  <Button variant="danger-invert" ml="2">
    Danger Invert
  </Button>
  <br />
  <br />
  <Button variant="regular">Regular</Button>
  <Button variant="normal" ml="2">
    Normal
  </Button>
</>
```

### Rounded

```jsx live
<>
  <Button rounded variant="primary">
    Primary
  </Button>
  <Button rounded variant="primary-invert" ml="2">
    Primary Invert
  </Button>
  <br />
  <br />
  <Button rounded variant="danger">
    Danger
  </Button>
  <Button rounded variant="danger-invert" ml="2">
    Danger Invert
  </Button>
  <br />
  <br />
  <Button rounded variant="regular">
    Regular
  </Button>
  <Button rounded variant="normal" ml="2">
    Normal
  </Button>
</>
```

### Icon

```js
import { MdArrowForward } from 'react-icons/md';
```

```jsx live
<>
  <Button icon={MdArrowForward} variant="primary">
    Primary
  </Button>
  <Button icon={MdArrowForward} variant="primary-invert" ml="2">
    Primary Invert
  </Button>
  <br />
  <br />
  <Button icon={MdArrowForward} variant="danger">
    Danger
  </Button>
  <Button icon={MdArrowForward} variant="danger-invert" ml="2">
    Danger Invert
  </Button>
  <br />
  <br />
  <Button icon={MdArrowForward} variant="regular">
    Regular
  </Button>
  <Button icon={MdArrowForward} variant="normal" ml="2">
    Normal
  </Button>
</>
```

```js
import { MdCheck } from 'react-icons/md';
```

```jsx live
<>
  <Button icon={MdCheck} variant="primary" rounded />
  <Button icon={MdCheck} variant="primary-invert" rounded ml="2" />
  <Button icon={MdCheck} variant="danger" rounded ml="2" />
  <Button icon={MdCheck} variant="danger-invert" rounded ml="2" />
  <Button icon={MdCheck} variant="regular" rounded ml="2" />
  <Button icon={MdCheck} variant="normal" rounded ml="2" />
</>
```

### Disabled

```jsx live
<>
  <Button disabled variant="primary">
    Primary
  </Button>
  <Button disabled variant="primary-invert" ml="2">
    Primary Invert
  </Button>
  <br />
  <br />
  <Button disabled variant="danger">
    Danger
  </Button>
  <Button disabled variant="danger-invert" ml="2">
    Danger Invert
  </Button>
  <br />
  <br />
  <Button disabled variant="regular">
    Regular
  </Button>
  <Button disabled variant="normal" ml="2">
    Normal
  </Button>
</>
```

### Loading

```jsx live
<>
  <Button loading variant="primary">
    Primary
  </Button>
  <Button loading variant="primary-invert" ml="2">
    Primary Invert
  </Button>
  <br />
  <br />
  <Button loading variant="danger">
    Danger
  </Button>
  <Button loading variant="danger-invert" ml="2">
    Danger Invert
  </Button>
  <br />
  <br />
  <Button loading variant="regular">
    Regular
  </Button>
  <Button loading variant="normal" ml="2">
    Normal
  </Button>
  <br />
  <br />
  <Button icon={MdCheck} loading ml="2" />
  <Button icon={MdCheck} loading ml="2" rounded />
</>
```

```jsx live
() => {
  const [loading, setLoading] = useState(false);

  return (
    <Button
      loading={loading}
      onClick={() => {
        setLoading(true);
        setTimeout(() => setLoading(false), 3000);
      }}
    >
      Click me!
    </Button>
  );
}
```

### Sizes

```jsx live
<>
  <Button size="sm">Button sm</Button>

  <Button ml={2}>Button md</Button>

  <Button ml={2} size="lg">
    Button lg
  </Button>
</>
```

## API

| Property   | Description                                             | Type                                                                               | Default |
|------------|---------------------------------------------------------|------------------------------------------------------------------------------------|---------|
| `variant`  | set different variant to display other colors of button | `'primary'` `'primary-invert'` `'danger'` `'danger-invert'` `'regular'` `'normal'` |         |
| `size`     | the size of button                                      | `'sm'` \| `'md'` \| `'lg'`                                                         | `'md'`  |
| `loading`  | set the loading status of button                        | `boolean`                                                                          | `false` |
| `rounded`  | set the button is rounded or not                        | `boolean`                                                                          | `false` |
| `disabled` | disabled state of button                                | `boolean`                                                                          | `false` |
| `icon`     | set the icon of button, see: [Icon](Icon.md) component  | `IconType`                                                                         |         |
| `onClick`  | set the handler to handle `click` event                 | `(event: MouseEvent) => void`                                                      |         |
