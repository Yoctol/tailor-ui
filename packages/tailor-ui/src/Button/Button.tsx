import React, { FunctionComponent, MouseEventHandler } from 'react';

import { IconType } from '../Icon';

import ButtonIcon from './ButtonIcon';
import { StyledButton, StyledButtonProps } from './styles';

export type ButtonProps = StyledButtonProps & {
  icon?: IconType;
  onClick?: MouseEventHandler;
};

const Button: FunctionComponent<ButtonProps> = ({
  children,
  icon,
  loading = false,
  size = 'md',
  ...props
}) => (
  <StyledButton
    hasIcon={Boolean(icon && !children)}
    size={size}
    loading={loading}
    {...props}
  >
    <ButtonIcon
      icon={icon}
      loading={loading}
      size={size}
      buttonContent={children}
    />
    {children}
  </StyledButton>
);

export default Button;
