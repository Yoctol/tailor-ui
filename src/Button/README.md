# Button Docs

```js
import { Button } from "@yoctol/ui";
```

```jsx
<Button>Text</Button>
```

Try it with `knobs`:

<!-- STORY -->

## API

| Props      | type                                                                              | Default | Description          |
| ---------- | --------------------------------------------------------------------------------- | ------- | -------------------- |
| `block`    | Boolean                                                                           | `false` |                      |
| `circle`   | Boolean                                                                           | `false` |                      |
| `active`   | Boolean                                                                           | `false` |                      |
| `ghost`    | Boolean                                                                           | `false` |                      |
| `light`    | Boolean                                                                           | `false` |                      |
| `disabled` | Boolean                                                                           | `false` |                      |
| `size`     | `'sm'`&#124;`'m'`&#124;`'lg'`                                                     | `'m'`   |                      |
| Space      | [View on docs](https://github.com/jxnblk/styled-system/blob/master/docs/table.md) |         | From `styled-stsyem` |
