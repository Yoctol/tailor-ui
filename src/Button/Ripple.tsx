import React, { MouseEvent, PureComponent, RefObject } from 'react';

import styled, { css, keyframes } from 'utils/styled-components';

const rippleAnimation = keyframes`
  100% {
    opacity: 0;
    transform: scale(3);
  }
`;

const StyledRipple = styled<{ rippling: boolean }, 'div'>('div')`
  position: absolute;
  border-radius: 50%;
  opacity: 1;
  background: rgba(0, 0, 0, 0.2);
  transform: scale(0);

  ${({ rippling }) =>
    rippling &&
    css`
      animation: ${rippleAnimation} 0.5s linear;
    `};
`;

class Ripple extends PureComponent {
  state = {
    animate: false,
    styles: {},
  };

  rippling = (
    { clientX, clientY }: { clientX: number; clientY: number },
    button: HTMLButtonElement
  ) => {
    const cursorPos = {
      top: clientY,
      left: clientX,
    };

    const buttonPos = button.getBoundingClientRect();
    const buttonWidth = button.offsetWidth;
    const buttonHeight = button.offsetHeight;

    const rippleWidth = Math.max(buttonHeight, buttonWidth);
    const rippleRadius = rippleWidth / 2;

    this.setState({
      animate: true,
      styles: {
        width: rippleWidth,
        height: rippleWidth,
        top: cursorPos.top - buttonPos.top - rippleRadius,
        left: cursorPos.left - buttonPos.left - rippleRadius,
      },
    });
  };

  startRipple = (
    { clientX, clientY }: MouseEvent,
    buttonRef: RefObject<any>
  ) => {
    const rippling = () => {
      this.rippling({ clientX, clientY }, buttonRef.current);
    };

    if (this.state.animate) {
      this.setState({ animate: false }, rippling);
    } else {
      rippling();
    }
  };

  render() {
    const { animate, styles } = this.state;
    return <StyledRipple rippling={animate} style={styles} />;
  }
}

export default Ripple;
