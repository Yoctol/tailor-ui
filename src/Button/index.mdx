---
name: Button
menu: General
route: /general/button
---

import { PropsTable, Playground } from 'docz';
import { MdCheck, MdDelete, MdClose, MdSearch } from 'react-icons/md';
import { Toggle } from 'react-powerplug';

import Box from '../Grid/Box';
import Icon from '../Icon';

import Button from './';

# Button

To trigger an operation.

## When To Use

A button means an operation (or a series of operations). Clicking a button will trigger corresponding business logic.

## Examples

```js
import { Button } from '@yoctol/ui';
```

### Basic

<Playground>
  <Button>Basic Button</Button>
  <Button text ml="2">Text Button</Button>
</Playground>

### Type

<Playground>
  <Button type="primary">Primary</Button>
  <Button type="secondary" ml="2">
    Secondary
  </Button>
  <Button type="info" ml="2">
    Info
  </Button>
  <Button type="success" ml="2">
    Success
  </Button>
  <Button type="warning" ml="2">
    Warning
  </Button>
  <Button type="danger" ml="2">
    Danger
  </Button>
</Playground>

### Rounded

<Playground>
  <Button type="primary" rounded>Primary</Button>
  <Button type="secondary" ml="2" rounded>
    Secondary
  </Button>
  <Button type="info" ml="2" rounded>
    Info
  </Button>
  <Button type="success" ml="2" rounded>
    Success
  </Button>
  <Button type="warning" ml="2" rounded>
    Warning
  </Button>
  <Button type="danger" ml="2" rounded>
    Danger
  </Button>
</Playground>

### Outlined

<Playground>
  <Button type="primary" outlined>Primary</Button>
  <Button type="secondary" ml="2" outlined>
    Secondary
  </Button>
  <Button type="info" ml="2" outlined>
    Info
  </Button>
  <Button type="success" ml="2" outlined>
    Success
  </Button>
  <Button type="warning" ml="2" outlined>
    Warning
  </Button>
  <Button type="danger" ml="2" outlined>
    Danger
  </Button>
</Playground>

### Rounded & Outlined

<Playground>
  <Button type="primary" rounded outlined>Primary</Button>
  <Button type="secondary" ml="2" rounded outlined>
    Secondary
  </Button>
  <Button type="info" ml="2" rounded outlined>
    Info
  </Button>
  <Button type="success" ml="2" rounded outlined>
    Success
  </Button>
  <Button type="warning" ml="2" rounded outlined>
    Warning
  </Button>
  <Button type="danger" ml="2" rounded outlined>
    Danger
  </Button>
</Playground>

### Icon

```js
import { MdCheck, MdDelete, MdClose, MdSearch } from 'react-icons/md';
```

<Playground>
  <Button type="primary" icon={MdCheck}>Confirm</Button>
  <Button type="danger" icon={MdDelete} ml="2">Delete</Button>
  <Button type="info" icon={MdSearch} ml="2">Search</Button>
  <br />
  <br />
  <Button type="primary" rounded icon={MdSearch} ml="2" />
  <Button type="danger" rounded icon={MdDelete} ml="2" />
  <Button type="info" rounded icon={MdSearch} ml="2" />
  <br />
  <br />
  <Button>
    <Icon type={MdCheck} fill="success" size="20" mr="2" />
    Success
  </Button>
  <Button ml="2">
    <Icon type={MdClose} fill="danger" size="20" mr="2" />
    Danger
  </Button>
</Playground>


### Disabled

<Playground>
  <Button type="primary" disabled>Primary</Button>
  <Button type="secondary" ml="2" disabled>
    Secondary
  </Button>
  <Button type="info" ml="2" disabled>
    Info
  </Button>
  <Button type="success" ml="2" disabled>
    Success
  </Button>
  <Button type="warning" ml="2" disabled>
    Warning
  </Button>
  <Button type="danger" ml="2" disabled>
    Danger
  </Button>
</Playground>

### Loading

<Playground>
  <Button type="primary" loading>Primary</Button>
  <Button type="secondary" ml="2" loading>
    Secondary
  </Button>
  <Button type="info" ml="2" loading>
    Info
  </Button>
  <Button type="success" ml="2" loading>
    Success
  </Button>
  <Button type="warning" ml="2" loading>
    Warning
  </Button>
  <Button type="danger" ml="2" loading>
    Danger
  </Button>
  <br />
  <br />
  <Toggle>
    {({ on, toggle }) => (
      <Button
        loading={on}
        onClick={() => {
          toggle()
          setTimeout(toggle, 3000)
        }}
      >
        Click me!
      </Button>
    )}
  </Toggle>
</Playground>

### Sizes

<Playground>
  <Button size="sm">
    Button sm
  </Button>

  <Button ml={2}>
    Button md
  </Button>

  <Button ml={2} size="lg">
    Button lg
  </Button>
</Playground>

### Block

<Playground>
  <Button block>
    Block Button
  </Button>
</Playground>

## API

<PropsTable of={Button} />
