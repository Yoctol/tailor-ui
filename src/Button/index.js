import PropTypes from 'prop-types';
import React, { PureComponent, createRef } from 'react';
import styled, { css, keyframes } from 'styled-components';
import { ifProp, switchProp } from 'styled-tools';
import { space, themeGet } from 'styled-system';

import controlTransition from '../utils/transition';

import Ripple from './Ripple';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

const StyledButton = styled.button`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border: ${themeGet('borders.default')};
  border-radius: ${themeGet('radii.1')};
  border-color: ${themeGet('colors.primary')};
  background-color: ${themeGet('colors.primaryDark')};
  color: ${themeGet('colors.light')};
  line-height: ${themeGet('lineHeight')};
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  &:focus {
    border-color: ${themeGet('colors.primaryDark')};
    outline: 0;
  }

  &:hover {
    background-color: ${themeGet('colors.light')};
    color: ${themeGet('colors.primaryDark')};
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  ${switchProp('size', {
    sm: css`
      padding: ${themeGet('space.paddingYSm')} ${themeGet('space.paddingXSm')};
      font-size: ${themeGet('fontSizes.sm')};
    `,
    m: css`
      padding: ${themeGet('space.paddingY')} ${themeGet('space.paddingX')};
      font-size: ${themeGet('fontSizes.default')};
    `,
    lg: css`
      padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};
      font-size: ${themeGet('fontSizes.lg')};
    `,
  })}

  ${ifProp(
    'block',
    css`
      width: 100%;
    `
  )}

  ${ifProp(
    'danger',
    css`
      border-color: ${themeGet('colors.error')};
      background-color: ${themeGet('colors.error')};
      color: ${themeGet('colors.light')};

      &:focus {
        border-color: ${themeGet('colors.error')};
      }

      &:hover {
        background-color: ${themeGet('colors.error')};
        color: ${themeGet('colors.light')};
      }
    `
  )};

  ${ifProp(
    'light',
    css`
      border-color: ${themeGet('colors.gray.8')};
      background-color: ${themeGet('colors.light')};
      color: ${themeGet('colors.gray.2')};

      &:focus {
        border-color: ${themeGet('colors.primaryDark')};
      }

      &:hover {
        background-color: ${themeGet('colors.gray.8')};
      }

      ${ifProp(
        'active',
        css`
          border-color: ${themeGet('colors.secondaryDark')};
          background-color: ${themeGet('colors.secondaryDark')};
          color: ${themeGet('colors.light')};

          &:hover {
            background-color: ${themeGet('colors.secondary')};
            color: ${themeGet('colors.light')};
          }

          &:focus {
            border-color: ${themeGet('colors.secondaryDark')};
          }
        `
      )};
    `
  )};

  ${ifProp(
    'circle',
    css`
    border-radius: 999px;

    ${ifProp(
      { size: 'sm' },
      css`
        padding: ${themeGet('space.paddingYSm')}
          calc(${themeGet('space.paddingXSm')} * 2);
      `
    )}
    ${ifProp(
      { size: 'm' },
      css`
        padding: ${themeGet('space.paddingY')}
          calc(${themeGet('space.paddingX')} * 2);
      `
    )}
    ${ifProp(
      { size: 'lg' },
      css`
        padding: ${themeGet('space.paddingYLg')}
          calc(${themeGet('space.paddingXLg')} * 2);
      `
    )}
  `
  )}

  ${ifProp(
    'ghost',
    css`
      border-color: ${themeGet('colors.light')};
      background-color: transparent;
    `
  )}

  ${ifProp(
    'loading',
    css`
      color: transparent;
      pointer-events: none;

      &::after {
        content: '';
        display: block;
        position: absolute;
        top: calc(50% - (1em / 2));
        left: calc(50% - (1em / 2));
        width: 1em;
        height: 1em;
        border: 2px solid ${themeGet('colors.gray.6')};
        border-radius: 50%;
        border-top-color: transparent;
        border-right-color: transparent;
        animation: ${spin} 0.7s linear infinite;
      }
    `
  )}

  ${controlTransition()};

  ${space};
`;

class Button extends PureComponent {
  constructor(props) {
    super(props);
    this.ripple = createRef();
    this.button = createRef();
  }

  handleClick = e => this.ripple.current.startRipple(e, this.button);

  render() {
    const { children, ...props } = this.props;
    return (
      <StyledButton
        innerRef={this.button}
        onMouseUp={this.handleClick}
        onTouchend={this.handleClick}
        {...props}
      >
        {children}
        <Ripple ref={this.ripple} />
      </StyledButton>
    );
  }
}

Button.displayName = 'Button';

Button.propTypes = {
  /**
   * Description of prop "active".
   */
  active: PropTypes.bool,
  /**
   * Description of prop "block".
   */
  block: PropTypes.bool,
  /**
   * Description of prop "circle".
   */
  circle: PropTypes.bool,
  /**
   * Description of prop "ghost".
   */
  danger: PropTypes.bool,
  /**
   * Description of prop "light".
   */
  ghost: PropTypes.bool,
  /**
   * Description of prop "danger".
   */
  light: PropTypes.bool,
  /**
   * Description of prop "loading".
   */
  loading: PropTypes.bool,
  /**
   * Description of prop "size".
   */
  size: PropTypes.oneOf(['sm', 'm', 'lg']),
  /**
   * Description of prop "Space".
   */
  ...space.propTypes,
};

Button.defaultProps = {
  size: 'm',
  block: false,
  circle: false,
  danger: false,
  ghost: false,
  light: false,
  active: false,
  loading: false,
};

export default Button;
