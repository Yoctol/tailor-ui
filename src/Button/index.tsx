import React, {
  MouseEvent,
  PureComponent,
  ReactInstance,
  RefObject,
  createRef,
} from 'react';
import { SpaceProps, space as styledSpace } from 'styled-system';

import styled, { css, keyframes } from 'utils/styled-components';
import tag from 'utils/CleanTag';

import Icon, { IconType, IconWrapper } from '../Icon';

import Ripple from './Ripple';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

type ButtonType =
  | 'default'
  | 'primary'
  | 'secondary'
  | 'info'
  | 'success'
  | 'warning'
  | 'danger';

export interface IGetTypesStylesInterface {
  type?: ButtonType;
  text?: boolean;
  outlined?: boolean;
}

const getTypeStyles = ({
  type = 'default',
  text,
  outlined,
  theme,
}: IGetTypesStylesInterface & {
  theme: any;
}) => {
  const { colors } = theme;

  if (text) {
    return {
      borderColor: 'transparent',
      backgroundColor: 'transparent',
      color: colors.gray800,
      hover: {
        backgroundColor: colors.gray300,
        borderColor: 'transparent',
      },
    };
  }

  if (type === 'default') {
    return {
      borderColor: colors.gray300,
      backgroundColor: colors.light,
      color: colors.gray800,
      hover: {
        backgroundColor: colors.gray200,
        borderColor: colors.primary,
      },
    };
  }

  const bg = colors[type];
  const { light } = colors;
  const backgroundColor = outlined ? 'transparent' : bg;

  return {
    borderColor: bg,
    backgroundColor,
    color: outlined ? bg : light,
    hover: {
      backgroundColor: outlined ? bg : light,
      color: outlined ? light : bg,
    },
  };
};

interface IGetLoadingInterface extends IGetTypesStylesInterface {
  loading?: boolean;
}

const getLoading = ({
  loading,
  ...props
}: IGetLoadingInterface & {
  theme: any;
}) =>
  loading &&
  css`
    opacity: 0.7;
    cursor: default;
    pointer-events: none;

    &::before {
      width: 1em;
      height: 1em;
      margin-right: 8px;
      border: 1px solid ${getTypeStyles(props).color};
      border-radius: 50%;
      border-top-color: transparent;
      border-right-color: transparent;
      animation: ${spin} 0.7s linear infinite;
    }
  `;

type ButtonSize = 'sm' | 'md' | 'lg';

interface IGetSizeInterface {
  size: ButtonSize;
  hasIcon?: boolean;
}

const getSize = css`
  ${({
    hasIcon,
    size,
    theme: { paddings, heights, fontSizes },
  }: IGetSizeInterface & { theme: any }) => {
    if (hasIcon) {
      return {
        sm: css`
          width: ${heights.sm};
          height: ${heights.sm};
        `,
        md: css`
          width: ${heights.base};
          height: ${heights.base};
        `,
        lg: css`
          width: ${heights.lg};
          height: ${heights.lg};
        `,
      }[size];
    }

    return {
      sm: css`
        height: ${heights.sm};
        padding: 0 ${paddings.sm};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 0 ${paddings.md};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 0 ${paddings.lg};
        font-size: ${fontSizes.lg};
      `,
    }[size];
  }};
`;

interface IBlockInterface {
  block?: boolean;
}

const getBlock = ({ block }: IBlockInterface) =>
  block &&
  css`
    width: 100%;
  `;

interface IRoundedInterface {
  rounded?: boolean;
}

const getRounded = ({ rounded }: IRoundedInterface) =>
  rounded &&
  css`
    border-radius: 999px;
  `;

const getTypes = ({
  type = 'default',
  text = false,
  outlined = false,
  theme,
}: IGetTypesStylesInterface & { theme: any }) => {
  const styles = getTypeStyles({ type, text, outlined, theme });

  return css`
    border-color: ${styles.borderColor};
    background-color: ${styles.backgroundColor};
    color: ${styles.color};

    ${IconWrapper /* sc-selector */} {
      cursor: pointer;
      svg {
        fill: ${styles.color};
      }
    }

    &:hover {
      border-color: ${styles.hover.borderColor};
      background-color: ${styles.hover.backgroundColor};
      color: ${styles.hover.color};

      ${IconWrapper /* sc-selector */} svg {
        fill: ${styles.hover.color};
      }
    }
  `;
};

type StyledButtonProps = IGetTypesStylesInterface &
  IGetLoadingInterface &
  IGetSizeInterface &
  IBlockInterface &
  IRoundedInterface &
  SpaceProps;

export const StyledButton = styled<StyledButtonProps, 'button'>(tag.button)`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  line-height: ${p => p.theme.lineHeight};
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  &:focus {
    outline: 0;
  }

  &:disabled,
  &[disabled] {
    opacity: 0.7;
    cursor: default;
    pointer-events: none;
  }

   &::before {
      content: '';
      display: inline-block;
      position: relative;
      left: 0;
      width: 0;
      height: 0;
      margin-right: 0;
      ${p => p.theme.transition /* sc-declaration */};
    }

  ${getBlock /* sc-declaration */}
  ${getRounded /* sc-declaration */}
  ${getSize /* sc-declaration */}
  ${getTypes /* sc-declaration */}
  ${getLoading /* sc-declaration */};

  ${p => p.theme.transition /* sc-declaration */};
  ${styledSpace};
`;

export type ButtonProps = SpaceProps & {
  type?: ButtonType;
  text?: boolean;
  block?: boolean;
  rounded?: boolean;
  outlined?: boolean;
  loading?: boolean;
  icon?: string | IconType;
  size?: ButtonSize;
  onClick?: (event: MouseEvent) => void;
};

class Button extends PureComponent<ButtonProps> {
  ripple: RefObject<Ripple> = createRef<Ripple>();

  buttonRef: ReactInstance | null = null;

  handleButtonRef = (ref: any) => {
    this.buttonRef = ref;
  };

  handleClick = (event: MouseEvent) => {
    if (this.ripple.current && this.buttonRef) {
      this.ripple.current.startRipple(event, this.buttonRef);
    }
  };

  render() {
    const {
      children,
      icon,
      loading = false,
      type = 'default',
      size = 'md',
      ...props
    } = this.props;

    return (
      <StyledButton
        ref={this.handleButtonRef}
        onMouseUp={this.handleClick}
        hasIcon={Boolean(icon && !children)}
        loading={loading}
        size={size}
        type={type}
        {...props}
      >
        {!loading &&
          icon && (
            <Icon
              type={icon}
              size="20"
              mr={children ? 1 : 0}
              style={{ pointerEvents: 'none' }}
            />
          )}
        {children}
        <Ripple ref={this.ripple} />
      </StyledButton>
    );
  }
}

export default Button;
