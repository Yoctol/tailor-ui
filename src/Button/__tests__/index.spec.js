import React from 'react';

import Button from '../';
import shallowWithTheme from '../../../test/shallowWithTheme';

describe('Button', () => {
  it('should render correctly', () => {
    const component = shallowWithTheme(<Button>Text Button</Button>);

    expect(component).toMatchSnapshot();
  });

  it('should render with props light, circle and active', () => {
    const component = shallowWithTheme(
      <Button light circle active>
        Text Button
      </Button>
    );

    expect(component).toMatchSnapshot();
  });
});
