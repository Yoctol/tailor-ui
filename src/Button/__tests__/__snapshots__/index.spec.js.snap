// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
.c1 {
  position: absolute;
  border-radius: 50%;
  opacity: 1;
  background: rgba(0,0,0,0.2);
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
  border: 0.0625rem solid;
  border-radius: 0.25rem;
  line-height: 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  padding: 0.3125rem 0.5rem;
  font-size: 1rem;
  border-color: #c3c3c3;
  background: #ffffff;
  color: #111111;
}

.c0:focus {
  outline: 0;
}

.c0:disabled,
.c0[disabled] {
  opacity: 0.5;
  cursor: default;
  pointer-events: none;
}

.c0:hover {
  border-color: #423b63;
  background: #e9e9e9;
}

<Button
  block={false}
  loading={false}
  outlined={false}
  rounded={false}
  size="md"
  text={false}
  type="default"
>
  <Button__StyledButton
    block={false}
    innerRef={
      Object {
        "current": <button
          class="c0"
          type="default"
        >
          Text Button
          <div
            class="c1"
          />
        </button>,
      }
    }
    loading={false}
    onMouseUp={[Function]}
    onTouchend={[Function]}
    outlined={false}
    rounded={false}
    size="md"
    text={false}
    type="default"
  >
    <button
      className="c0"
      onMouseUp={[Function]}
      size="md"
      type="default"
    >
      Text Button
      <Ripple>
        <Ripple__StyledRipple
          rippling={false}
          style={Object {}}
        >
          <div
            className="c1"
            style={Object {}}
          />
        </Ripple__StyledRipple>
      </Ripple>
    </button>
  </Button__StyledButton>
</Button>
`;

exports[`Button should render with props variant and type 1`] = `
.c1 {
  position: absolute;
  border-radius: 50%;
  opacity: 1;
  background: rgba(0,0,0,0.2);
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
  border: 0.0625rem solid;
  border-radius: 0.25rem;
  line-height: 1.5rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  padding: 0.3125rem 0.5rem;
  font-size: 1rem;
  border-color: #4384f0;
  background: #4384f0;
  color: #fff;
}

.c0:focus {
  outline: 0;
}

.c0:disabled,
.c0[disabled] {
  opacity: 0.5;
  cursor: default;
  pointer-events: none;
}

.c0:hover {
  background: #fff;
  color: #4384f0;
}

<Button
  block={false}
  loading={false}
  outlined={false}
  rounded={false}
  size="md"
  text={false}
  type="info"
  variant="rounded"
>
  <Button__StyledButton
    block={false}
    innerRef={
      Object {
        "current": <button
          class="c0"
          type="info"
        >
          Text Button
          <div
            class="c1"
          />
        </button>,
      }
    }
    loading={false}
    onMouseUp={[Function]}
    onTouchend={[Function]}
    outlined={false}
    rounded={false}
    size="md"
    text={false}
    type="info"
    variant="rounded"
  >
    <button
      className="c0"
      onMouseUp={[Function]}
      size="md"
      type="info"
    >
      Text Button
      <Ripple>
        <Ripple__StyledRipple
          rippling={false}
          style={Object {}}
        >
          <div
            className="c1"
            style={Object {}}
          />
        </Ripple__StyledRipple>
      </Ripple>
    </button>
  </Button__StyledButton>
</Button>
`;
